core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/common/WorkflowActor.scala

```diff
@@ -1,28 +1,46 @@
 package edu.uci.ics.amber.engine.architecture.common
 
-import akka.actor.{Actor, ActorLogging, Stash}
+import akka.actor.{Actor, ActorLogging, ActorRef, Stash}
 import com.softwaremill.macwire.wire
+import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkSenderActor.{
+  NetworkSenderActorRef,
+  QueryActorRef,
+  RegisterActorRef
+}
 import edu.uci.ics.amber.engine.architecture.messaginglayer.{
   ControlInputPort,
   ControlOutputPort,
-  NetworkOutputGate
+  NetworkSenderActor
 }
-import edu.uci.ics.amber.engine.common.ambertag.neo.VirtualIdentity
+import edu.uci.ics.amber.engine.common.amberexception.WorkflowRuntimeException
 import edu.uci.ics.amber.engine.common.ambertag.neo.VirtualIdentity.ActorVirtualIdentity
+import edu.uci.ics.amber.error.WorkflowRuntimeError
 
-class WorkflowActor(identifier: ActorVirtualIdentity)
+abstract class WorkflowActor(identifier: ActorVirtualIdentity)
     extends Actor
     with ActorLogging
-    with Stash
-    with NetworkOutputGate {
+    with Stash {
 
-  val networkOutput: NetworkOutputGate = this
+  val networkSenderActor: NetworkSenderActorRef = NetworkSenderActorRef(
+    context.actorOf(NetworkSenderActor.props())
+  )
   lazy val controlInputPort: ControlInputPort = wire[ControlInputPort]
   lazy val controlOutputPort: ControlOutputPort = wire[ControlOutputPort]
 
-  // Not being used right now.
-  override def receive: Receive = {
-    case other =>
-      throw new NotImplementedError("Message other than data message reached WorkflowActor receive")
+  def routeActorRefRelatedMessages: Receive = {
+    case QueryActorRef(id, replyTo) =>
+      if (replyTo.contains(networkSenderActor.ref)) {
+        context.parent ! QueryActorRef(id, replyTo)
+      } else {
+        networkSenderActor ! QueryActorRef(id, replyTo)
+      }
+    case RegisterActorRef(id, ref) =>
+      throw WorkflowRuntimeException(
+        WorkflowRuntimeError(
+          "workflow actor should never receive register actor ref message",
+          identifier.toString,
+          Map.empty
+        )
+      )
```

Comment:
Also mention that there is no handling of replyActorRef here because only network actors query actor refs. In fact, we should raise a WorkflowRuntimeException if replyActorRef is received.

Reply:
I already did that. replyActorRef is renamed to RegisterActorRef